klippy/extras/dockable_probe.py

```diff
@@ -0,0 +1,596 @@
+            self.toolhead.manual_move(
+                [self.detach_position[0], self.detach_position[1], None],
+                 self.travel_speed)
+        if self.get_probe_state() != PROBE_DOCKED:
+            raise self.printer.command_error('Probe detach failed!')
+        self.post_detach_gcode.run_gcode_from_command()
+        if return_pos:
+            if not self._check_distance(return_pos, self.detach_distance):
+                self.toolhead.manual_move(
+                    [return_pos[0], return_pos[1], None],
+                    self.travel_speed)
+    def _do_detach(self):
+        if self._check_distance(dist=self.detach_distance):
+            self._align_to_vector(self.dock_angle)
+        else:
+            self.toolhead.manual_move(
+                [self.detach_position[0], self.detach_position[1], None],
+                 self.travel_speed)
+        self.toolhead.manual_move(
+            [self.dock_position[0], self.dock_position[1], None],
+             self.attach_speed)
+        self.detach_gcode.run_gcode_from_command()
+        self.toolhead.manual_move(
+            [self.detach_position[0], self.detach_position[1], None],
+             self.travel_speed)
+
+    # Register a callback to detach the probe in the future if
+    # no pending probing actions are queued
+    def _delayed_detach(self):
+        if self.init_pos:
+            return_pos = self.init_pos
+        else:
+            return_pos = self.toolhead.get_position()
+        def reactor_bgfunc(print_time):
+            if self.multi == MULTI_OFF:
+                with self.gcode.get_mutex():
+                    self.detach_probe(return_pos)
+        def lookahead_bgfunc(print_time):
+            reactor = self.printer.get_reactor()
+            reactor.register_callback(
+                ( lambda et: reactor_bgfunc(print_time + PROBE_DETACH_DELAY )))
+        self.toolhead.register_lookahead_callback(lookahead_bgfunc)
+
+    #######################################################################
+    # Events handlers
+    #######################################################################
+    def _handle_homing_rails_begin(self, homing_state, rails):
+        homing_axes = [a for i, a in enumerate(['x','y','z'])
+                        if i in homing_state.get_axes()]
+        if self in [es for rail in rails for es, name in rail.get_endstops()]:
+            self.is_z_endstop = True
+
+        # Get homed axes now as they will report they are
+        # homed later even if they arent
+        curtime = self.printer.get_reactor().monotonic()
+        self._last_homed = self.toolhead.get_status(curtime)['homed_axes']
+        if self.get_probe_state() == PROBE_UNKNOWN:
+                self._force_z_hop()
+        if self.get_probe_state() == PROBE_ATTACHED:
+            # If X and Y endstop are within the dock location,
+            # Move out of the dock before attempting to home Z
+            if any(a in ['x','y'] for a in homing_axes):
```

Comment:
Maybe `if 'x' in homing_axes or 'y' in homing_axes:` is easier to read?

Reply:
thats how I had it written before then decided it was more illegible this way.